### DISCUSSION TITLE
num_items_and_pages() 3-4x slower than normal query

### DISCUSSION THREAD

**Discussion by rakbladsvalsen** [SeaQL/sea-orm#1928] [General]
## Description

The num_items_and_pages() method appears to be significantly slower than a raw query, for some weird reason. 

## Steps to Reproduce

1. Create a table with the schema (see Reproducible Example)
2. Perform a raw query to get the total count
3. Perform a query using exclusively SeaORM + num_items_and_pages()

### Expected Behavior

Both queries should take roughly the same time to complete.

### Actual Behavior

SeaORM's num_items_and_pages() is significantly slower, taking about 4x-5x more time with 1M entries.

### Reproduces How Often

100% reproducible

### Workarounds

N/A

## Reproducible Example

Schema:

- id (bigint, autoincrement)
- 1 foreign key (in this example `format_id`)
- 1 JSONB column (in this example, it's `data`)

</div><br class="Apple-interchange-newline">
JSONB data example:

```
{"SomeColumn": 1234, "SomeOtherColumn": "textData"},
{"SomeColumn": 1235, "SomeOtherColumn": "textData2"}
...
```
** IMPORTANT NOTE:** The difference gets noticeably bigger when you have at least 1M entries. Otherwise, the time difference is minimal at best.

```
  // Raw Query
    let query_two = r#"
    SELECT
    COUNT(*) AS num_items
  FROM
    (
      SELECT
        "record"."id",
        "record"."format_id",
        "record"."data"
      FROM
        "record"
      WHERE
        "record"."format_id" = $1
        AND CAST(("data" ->> $2) AS FLOAT) >= $3
    ) AS "sub_query"
    "#;

    let result = sea_orm::JsonValue::find_by_statement(Statement::from_sql_and_values(
        DbBackend::Postgres,
        query_two,
        [220.into(), "SomeColumn".into(), 1234.into()],
    ))
    .all(db)
    .await?;

    info!("test output: {:?}", result);

    // Using SeaORM's 
    let expr = Expr::col(entity::record::Column::Data)
        .binary(PgBinOper::CastJsonField, "SomeColumn")
        .cast_as(Alias::new("FLOAT"))
        .eq(1234);
    let entries = entity::record::Entity::find()
        .filter(entity::record::Column::FormatId.eq(220))
        .filter(expr)
        .paginate(db, 20000);
    let result = entries.num_items_and_pages().await?;
    info!("test output 2: {:?}", result);
```
Ironically, the raw query I used is basically the same thing that SeaORM is using to pull the total item count, so I don't know what is really going on. There must be something in SeaORM causing Postgres to be unbelievably slow, but I couldn't pinpoint it. 
I'm attaching a wireshark packet capture anyway. The only difference I noticed is that Postgres returns a "Portal-Suspend" 
type in the result when I use SeaORM, but it doesn't return that with the raw query. I didn't notice anything else significantly different.

## Versions

Linux 6.x
Rustc: rustc 1.72.1 (d5c2e9c34 2023-09-13)

```
cargo tree | grep -i sea-orm
│   ├── sea-orm v0.12.3
│   │   ├── sea-orm-macros v0.12.3 (proc-macro)
├── sea-orm v0.12.3 (*)

cargo tree | grep -i sql* 
│   │   │   ├── sqlx v0.7.2
│   │   │   │   ├── sqlx-core v0.7.2
│   │   │   │   │   ├── sqlformat v0.2.2
│   │   │   │   └── sqlx-postgres v0.7.2
│   │   │   │       ├── sqlx-core v0.7
```

**Comment by rakbladsvalsen** [comment]
Here's the wireshark packet capture:

Packet 13: "raw" query
14: Result response

Packet 21: Query built using SeaORM
Packet 22: Result response

You can notice the time difference between both. 

[packets.zip](https://github.com/SeaQL/sea-orm/files/12772888/packets.zip)

**Comment by tyt2y3** [comment]
I don't have a firm idea either, but I would have guessed it's something about the prepared query cache.

**Comment by rakbladsvalsen** [comment]
> I don't have a firm idea either, but I would have guessed it's something about the prepared query cache.

Hey @tyt2y3, thanks for the reply! 

I was able to pinpoint the exact issue. For some super weird reason, the method `query_one()` inside `num_items_and_pages()` makes everything 2-3x slower. I rolled out my own implementation of num_items_and_pages() using `query_all` instead:

```
    async fn num_items_and_pages<C: ConnectionTrait>(
        db: &C,
        select: &mut sea_orm::Select<Self::Entity>,
        page_size: u64,
    ) -> Result<(u64, u64), DbErr> {
        let select = SelectStatement::new()
            .expr(Expr::cust("COUNT (*) AS num_items"))
            .from_subquery(
                sea_orm::QueryTrait::query(select).to_owned(),
                Alias::new("sub_query"),
            )
            .to_owned();

        let stmt = StatementBuilder::build(&select, &sea_orm::DatabaseBackend::Postgres);

        let result = db.query_all(stmt).await?;
        let result = match result.get(0) {
            Some(i) => i,
            _ => return Ok((0, 0)),
        };
        let num_items = result.try_get::<u64>("", "num_items")?;
        let num_pages = (num_items as f64 / page_size as f64).ceil() as u64;
        Ok((num_items, num_pages))
    }
```

And the time taken to return the count is in line with what I'm always expecting (and just as fast as JDBC or psycopg2). Do you know what could be possibly causing this? I can make a PR for this, but I feel like this is more of a band-aid fix for something wrong lying deep within `query_one()`.

**Comment by tyt2y3** [comment]
Oh, then now it sounds like a query plan issue. The only practical difference between them is query one does `LIMIT 1`.

**Comment by rakbladsvalsen** [comment]
> Oh, then now it sounds like a query plan issue. The only practical difference between them is query one does `LIMIT 1`.

That's weird.  I tried doing a `COUNT` using adminer (uses PHP), dbeaver (JDBC) and python (psycopg2) with a `LIMIT 1` statement at the end and it completes in less than 1s. I see the same behavior if I use SeaORM's `query_all()`. 

There's probably something else (other than the `LIMIT `) in SeaORM that causes Postgres to be slow. It doesn't appear to be the query builder because queries are built in less than 1ms, and the slowdown comes after SeaORM sends the SQL query to the Postgres server. 

I absolutely cannot blame Postgres either because it works fine with any other ORM, using the same count query. I'm using the `latest` image from dockerhub BTW.

**Comment by tyt2y3** [comment]
That's all I can comment given the information at hand
